### PR TITLE
Add TypedStage to handle typed stages

### LINKED ISSUE #139: Implement typed stages
We need to be able to store typed stages like "DTR" (DTS, DPAS?) in identifier's object.
And provide "fuzzy" harmonized stages related to this "typed stage" when there are several options available.

From Skype chat with @ronaldtse:
> Maybe instead of a new class `PossibleHarmonizedStages`, you can change `HarmonizedStage` into a "fuzzy" specification, e.g. `HarmonizedStage.new(:fuzzy, "10.00", "20.00"...)` ?

> For a fuzzy `HarmonizedStage` that goes across stages (e.g. "10.99", "20.00") will be undetermined. But if the `HarmonizedStage` is `"10.xx"` then we have the stage code as 10, just no substage.

> So you are right... fuzzy HarmonizedStage should not allow generating URNs

> So there are only 3 statuses possible:
> - published (i.e. don't specify `status` at all)
> - draft (i.e. stages from 00.xx to 50.xx)
> - cancelled (i.e. stages from 00.98, 10.98, 20.98, 30.98, 40.98, 50.98, 95.99)
> so i guess if a HarmonizedStage code is 00.xx to 50.xx, we can use the `draft` value instead of `stage-xx.yy`.
> 

"DTR" is the type-stage-abbreviation of "TR" in stage 40 to 50. Notice that stage code 30 is a "CD TR".

Perhaps this is useful as a clarification of the type-stage definitions. i.e.

```ruby
TECHNICAL_REPORT = {
  type_stage: {
    pwitr: {
      abbr: "PWI TR",
      name: "Preliminary Work Item Technical Report",
      harmonized_stages: [ "10.00", "10.20", ... ]
    },
    awitr: {
      abbr: "AWI TR",
      name: "Approved Work Item Technical Report",
      harmonized_stages: [ "10.99" ]
    },
    wdtr: {
      abbr: "WD TR",
      name: "Working Draft Technical Report",
      harmonized_stages: [ 20 ]
    },
    wdtr: {
      abbr: "WD TR",
      name: "Working Draft Technical Report",
      harmonized_stages: [ 20 ]
    },
    cdtr: {
      abbr: "CD TR",
      name: "Committee Draft Technical Report",
      harmonized_stages: [ 30 ]
    },
    dtr: {
      abbr: "DTR",
      name: "Draft Technical Report",
      harmonized_stages: [ 40, 50 ]
    },
    tr: {
      abbr: "TR",
      name: "Technical Report",
      harmonized_stages: [ "60" ] # etc.
  }
}
```

We can define this for every PubID document type.

_Originally posted by @ronaldtse in https://github.com/metanorma/pubid-iso/pull/134#discussion_r994780449_

### PR DESCRIPTION
**Pull request by mico**
closes #139

### CODE REVIEW

**Comment by ronaldtse** [comment]
As per https://github.com/metanorma/pubid-iso/pull/134#discussion_r997664855 , we should move the TypedStage definitions inside:
* `Identifier::InternationalStandard`
* `Identifier::TechnicalReport`
* etc.

So that it is easier to manage future changes of these identifiers.

**Review by ronaldtse — CHANGES_REQUESTED** [review]
This needs to be applied to all document types, and also the TypedStage contents extracted into separate Identifier classes.

File: lib/pubid/iso/harmonized_stage_code.rb
Comment (ronaldtse) [inline lib/pubid/iso/harmonized_stage_code.rb:60]: 90.99 is "Withdrawal of International Standard proposed by TC or SC", it is a "proposal" to withdraw, it has not been withdrawn. So it is still in published stage.

This is the only one to move to PUBLISHED_STAGES.

File: spec/pubid_iso/harmonized_stage_code_spec.rb
Comment (ronaldtse) [inline spec/pubid_iso/harmonized_stage_code_spec.rb:29]: Excellent!

File: lib/pubid/iso/harmonized_stage_code.rb
Comment (ronaldtse) [inline lib/pubid/iso/harmonized_stage_code.rb:125]: You may also like to add the `<>` method to allow checking if a "00.00" is within the "00" stage, etc.

File: lib/pubid/iso/harmonized_stage_code.rb
Comment (mico) [reply lib/pubid/iso/harmonized_stage_code.rb:130]: @ronaldtse we can have many stages (stage and substage together).
But in case if we have more than one stage we output status (i.e., "draft" or "canceled" or "published") when all stages within one status, if not, we raise an error.

File: lib/pubid/iso/harmonized_stage_code.rb
Comment (mico) [reply lib/pubid/iso/harmonized_stage_code.rb:125]: > You may also like to add the `<>` method to allow checking if a "00.00" is within the "00" stage, etc.

Now there is no way to compare it without substage. How do you imagine it?

Comment (ronaldtse) [reply lib/pubid/iso/harmonized_stage_code.rb:125]: I was thinking of sorting by stages and then substages within stages. But clearly it's not important here and shouldn't block the merge.

File: lib/pubid/iso/harmonized_stage_code.rb
Comment (ronaldtse) [reply lib/pubid/iso/harmonized_stage_code.rb:130]: OK, thanks!

**Comment by ronaldtse** [comment]
@mico can we merge this once the tests pass? Thanks.

File: lib/pubid/iso/amendment.rb
Comment (ronaldtse) [inline lib/pubid/iso/amendment.rb:8]: Will this end up with a slash followed by an empty space? It seems that some defensive programming is necessary here.

File: lib/pubid/iso/corrigendum.rb
Comment (ronaldtse) [inline lib/pubid/iso/corrigendum.rb:6]: As above.

File: lib/pubid/iso/typed_stage.rb
Comment (ronaldtse) [inline lib/pubid/iso/typed_stage.rb:6]: Later on let’s split this information into individual identifier classes. We have information splattered everywhere about different document types.

Comment (ronaldtse) [reply lib/pubid/iso/typed_stage.rb:6]: I think we should really incorporate the Excel table somehow, in each identifier class.

**Review by ronaldtse — CHANGES_REQUESTED** [review]

File: lib/pubid/iso/amendment.rb
Comment (mico) [reply lib/pubid/iso/amendment.rb:8]: > Will this end up with a slash followed by an empty space? It seems that some defensive programming is necessary here.

@typed_stage.to_s will always return something. Without stage or typed stage it will return "Amd" for amendment.
Actually, we don't need separate classes for Amendment and Corrigendum, it's legacy now.

**Comment by ronaldtse** [comment]
Can we merge this now?

**Comment by ronaldtse** [comment]
This is taking way too long. Merging this first.

**Review by ronaldtse — APPROVED** [review]

File: lib/pubid/iso/amendment.rb
Comment (ronaldtse) [reply lib/pubid/iso/amendment.rb:8]: What if this is an "IS"? There is no Typed Stage abbreviation and there should not be any slash, see #151

**Pull request merged**
